### ISSUE TITLE
Configuration caching - serializing file-based dependencies may fail

### ISSUE THREAD

**Issue by gavra0** [gradle/gradle#14513]
Using configuration caching with the example below fails:
```
dependencies {
    compile project(":lib")
}
import org.gradle.api.artifacts.transform.*
/** Consumer task */
abstract class MyTask extends DefaultTask {
    @InputFiles
    abstract ConfigurableFileCollection getMyTaskInputs();
    @OutputFile
    abstract RegularFileProperty getMyTaskOutputFile();
    @TaskAction
    public void processing() {
        File file = getMyTaskOutputFile().get().asFile
        file.parentFile.mkdirs()
        file.delete()
        file.write(getMyTaskInputs().files.join(";\n"))
    }
}
def artifactType = Attribute.of("artifactType", String.class)
FileCollection sourceTransformed = project.files(configurations.getByName("compile").incoming.artifactView { viewConfig ->
    viewConfig.attributes.attribute(artifactType, "intermediate")
}.files)
Configuration configurationTransformed = project.configurations.detachedConfiguration(project.dependencies.create(sourceTransformed))
abstract class MyTransform implements TransformAction<TransformParameters.None> {
    @InputArtifact
    public abstract Provider<FileSystemLocation> getInputArtifact();
    @Override
    void transform(TransformOutputs transformOutputs) {
        File input = getInputArtifact().get().asFile
        if (input.exists()) {
            java.nio.file.Files.copy(input.toPath(), transformOutputs.file(input.name).toPath())
        } else {
            throw new RuntimeException("File does not exist: " + input)
        }
    }
}
project.dependencies.registerTransform(MyTransform.class) { transformSpec ->
    transformSpec.from.attribute(artifactType, "jar")
    transformSpec.to.attribute(artifactType, "intermediate")
}
project.dependencies.registerTransform(MyTransform.class) { transformSpec ->
    transformSpec.from.attribute(artifactType, "directory")
    transformSpec.to.attribute(artifactType, "intermediate")
}
project.dependencies.registerTransform(MyTransform.class) { transformSpec ->
    transformSpec.from.attribute(artifactType, "intermediate")
    transformSpec.to.attribute(artifactType, "my-final-type")
}
project.dependencies.registerTransform(MyTransform.class) { transformSpec ->
    transformSpec.from.attribute(artifactType, "intermediate")
    transformSpec.to.attribute(artifactType, "my-final-type")
}
project.tasks.register("myTask", MyTask.class) {
    it.getMyTaskInputs().from(
            configurationTransformed.incoming.artifactView { viewConfig ->
                viewConfig.attributes.attribute(artifactType, "my-final-type")
            }.files
    )
    it.getMyTaskOutputFile().fileValue(new File(project.buildDir, "myTask/output.txt"))
}
```

Runnning `./gradlew myTask --configuration-cache` will fail because jar produced by `:lib` is missing when task graph is serialized.

The root cause seems to be https://github.com/gradle/gradle/blob/5aca1dc8e4cdafb17ee7cf9cabe680c7ba93c1e6/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/LocalFileDependencyBackedArtifactSet.java#L83 where source of the transforms file collection will be resolved, even if it requires running artifact transforms.

**Comment by adammurdoch** [comment]
There is an initial fix for this available on `master` now. We might make some more changes depending on the performance impact of the fix.

**Comment by eskatos** [comment]
This is fixed in 6.8, closing
